diff --git a/src/image.c b/src/image.c
index fa8b0ea..ccac7a6 100644
--- a/src/image.c
+++ b/src/image.c
@@ -155,17 +155,9 @@ static VAStatus copy_surface_to_image (struct request_data *driver_data,
 		return VA_STATUS_ERROR_INVALID_BUFFER;
 
 	for (i = 0; i < surface_object->destination_planes_count; i++) {
-		if (!video_format_is_linear(driver_data->video_format))
-			tiled_to_planar(surface_object->destination_data[i],
-					buffer_object->data + image->offsets[i],
-					image->pitches[i], image->width,
-					i == 0 ? image->height :
-						 image->height / 2);
-		else {
-			memcpy(buffer_object->data + image->offsets[i],
-			       surface_object->destination_data[i],
-			       surface_object->destination_sizes[i]);
-		}
+		memcpy(buffer_object->data + image->offsets[i],
+		       surface_object->destination_data[i],
+		       surface_object->destination_sizes[i]);
 	}
 
 	return VA_STATUS_SUCCESS;
