diff --git a/arch/arm/boot/dts/sun8i-a83t.dtsi b/arch/arm/boot/dts/sun8i-a83t.dtsi
index a6fd8450c8bc..51053f83b386 100644
--- a/arch/arm/boot/dts/sun8i-a83t.dtsi
+++ b/arch/arm/boot/dts/sun8i-a83t.dtsi
@@ -215,24 +215,6 @@ opp-1008000000 {
 			opp-microvolt = <840000>;
 			clock-latency-ns = <244144>; /* 8 32k periods */
 		};
-
-		opp-1412000000 {
-			opp-hz = /bits/ 64 <1412000000>;
-			opp-microvolt = <920000>;
-			clock-latency-ns = <244144>; /* 8 32k periods */
-		};
-
-		opp-1608000000 {
-			opp-hz = /bits/ 64 <1608000000>;
-			opp-microvolt = <1000000>;
-			clock-latency-ns = <244144>; /* 8 32k periods */
-		};
-
-		opp-1800000000 {
-			opp-hz = /bits/ 64 <1800000000>;
-			opp-microvolt = <1080000>;
-			clock-latency-ns = <244144>; /* 8 32k periods */
-		};
 	};
 
 	cpu1_opp_table: opp_table1 {
@@ -250,24 +232,6 @@ opp-1008000000 {
 			opp-microvolt = <840000>;
 			clock-latency-ns = <244144>; /* 8 32k periods */
 		};
-
-		opp-1412000000 {
-			opp-hz = /bits/ 64 <1412000000>;
-			opp-microvolt = <920000>;
-			clock-latency-ns = <244144>; /* 8 32k periods */
-		};
-
-		opp-1608000000 {
-			opp-hz = /bits/ 64 <1608000000>;
-			opp-microvolt = <1000000>;
-			clock-latency-ns = <244144>; /* 8 32k periods */
-		};
-
-		opp-1800000000 {
-			opp-hz = /bits/ 64 <1800000000>;
-			opp-microvolt = <1080000>;
-			clock-latency-ns = <244144>; /* 8 32k periods */
-		};
 	};
 
 	scpi_protocol: scpi {
