diff --git a/arch/arm64/boot/dts/amlogic/Makefile b/arch/arm64/boot/dts/amlogic/Makefile
index 73d96042d570..3957a5be9022 100644
--- a/arch/arm64/boot/dts/amlogic/Makefile
+++ b/arch/arm64/boot/dts/amlogic/Makefile
@@ -26,10 +26,6 @@ dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905x-khadas-vim.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905x-libretech-cc.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905x-nexbox-a95x.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905x-p212.dtb
-dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905d-mecool-ki-plus.dts
-dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905d-mecool-ki-pro.dts
-dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905d-mecool-kii-pro.dts
-dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905d-mecool-m8s-plus.dts
 dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905d-p230.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905d-p231.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-gxl-s905d-phicomm-n1.dtb
diff --git a/arch/arm64/boot/dts/amlogic/meson-gxbb-mecool-ki-plus.dts b/arch/arm64/boot/dts/amlogic/meson-gxbb-mecool-ki-plus.dts
deleted file mode 100644
index f08e2a7178c5..000000000000
--- a/arch/arm64/boot/dts/amlogic/meson-gxbb-mecool-ki-plus.dts
+++ /dev/null
@@ -1,34 +0,0 @@
-// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
-
-/*
- * Author: Christian Hewitt <christianshewitt@gmail.com>
- */
-
-#include "meson-gxbb-p201.dts"
-
-/ {
-        compatible = "mecool,gxbb-ki-plus", "amlogic,meson-gxbb";
-	model = "MeCool KI Plus";
-
-	clock: meson_clock {
-		compatible = "amlogic, gxbb-clock";
-		reg = <0x0 0xc883c000 0x0 0x1000>,
-			<0x0 0xc8100000 0x0 0x1000>;
-		#clock-cells = <1>;
-		#reset-cells = <1>;
-		sys_max = <1536000000>;
-	};
-
-	memory@0 {
-		device_type = "memory";
-		reg = <0x0 0x0 0x0 0x40000000>;
-	};
-};
-
-&ir {
-        linux,rc-map-name = "rc-mecool-ki-plus";
-};
-
-&usb_pwr {
-	gpio = <>;
-};
diff --git a/arch/arm64/boot/dts/amlogic/meson-gxbb-mecool-kii-pro.dts b/arch/arm64/boot/dts/amlogic/meson-gxbb-mecool-kii-pro.dts
deleted file mode 100644
index 4f43aff680c3..000000000000
--- a/arch/arm64/boot/dts/amlogic/meson-gxbb-mecool-kii-pro.dts
+++ /dev/null
@@ -1,34 +0,0 @@
-// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
-
-/*
- * Author: Christian Hewitt <christianshewitt@gmail.com>
- */
-
-#include "meson-gxbb-p201.dts"
-
-/ {
-        compatible = "mecool,gxbb-kii-pro", "amlogic,meson-gxbb";
-	model = "MeCool KII Pro";
-
-	clock: meson_clock {
-		compatible = "amlogic, gxbb-clock";
-		reg = <0x0 0xc883c000 0x0 0x1000>,
-			<0x0 0xc8100000 0x0 0x1000>;
-		#clock-cells = <1>;
-		#reset-cells = <1>;
-		sys_max = <1536000000>;
-	};
-
-	memory@0 {
-		device_type = "memory";
-		reg = <0x0 0x0 0x0 0x40000000>;
-	};
-};
-
-&ir {
-        linux,rc-map-name = "rc-mecool-kii-pro";
-};
-
-&usb_pwr {
-	gpio = <>;
-};
diff --git a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-ki-plus.dts b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-ki-plus.dts
deleted file mode 100644
index c034225b9d9c..000000000000
--- a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-ki-plus.dts
+++ /dev/null
@@ -1,21 +0,0 @@
-// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
-
-/*
- * Author: Christian Hewitt <christianshewitt@gmail.com>
- */
-
-#include "meson-gxl-s905d-p231.dts"
-
-/ {
-        compatible = "mecool,ki-plus", "amlogic,s905d", "amlogic,meson-gxl";
-	model = "MeCool KI Plus";
-
-	memory@0 {
-		device_type = "memory";
-		reg = <0x0 0x0 0x0 0x40000000>;
-	};
-};
-
-&ir {
-        linux,rc-map-name = "rc-mecool-ki-plus";
-};
diff --git a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-ki-pro.dts b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-ki-pro.dts
deleted file mode 100644
index 606c5dc2efb2..000000000000
--- a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-ki-pro.dts
+++ /dev/null
@@ -1,16 +0,0 @@
-// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
-
-/*
- * Author: Christian Hewitt <christianshewitt@gmail.com>
- */
-
-#include "meson-gxl-s905d-p230.dts"
-
-/ {
-        compatible = "mecool,ki-pro", "amlogic,s905d", "amlogic,meson-gxl";
-	model = "MeCool KI Pro";
-};
-
-&ir {
-        linux,rc-map-name = "rc-mecool-ki-pro";
-};
diff --git a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-kii-pro.dts b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-kii-pro.dts
deleted file mode 100644
index 8db4d097f667..000000000000
--- a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-kii-pro.dts
+++ /dev/null
@@ -1,16 +0,0 @@
-// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
-
-/*
- * Author: Christian Hewitt <christianshewitt@gmail.com>
- */
-
-#include "meson-gxl-s905d-p231.dts"
-
-/ {
-        compatible = "mecool,kii-pro", "amlogic,s905d", "amlogic,meson-gxl";
-	model = "MeCool KII Pro";
-};
-
-&ir {
-        linux,rc-map-name = "rc-mecool-kii-pro";
-};
diff --git a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-m8s-plus.dts b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-m8s-plus.dts
deleted file mode 100644
index 2da13c9f69b6..000000000000
--- a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-mecool-m8s-plus.dts
+++ /dev/null
@@ -1,16 +0,0 @@
-// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
-
-/*
- * Author: Christian Hewitt <christianshewitt@gmail.com>
- */
-
-#include "meson-gxl-s905d-p231.dts"
-
-/ {
-        compatible = "mecool,kii-pro", "amlogic,s905d", "amlogic,meson-gxl";
-	model = "MeCool M8S Plus";
-};
-
-&ir {
-        linux,rc-map-name = "rc-mecool-m8s-plus";
-};
diff --git a/arch/arm64/boot/dts/amlogic/meson-gxm-mecool-kiii-pro.dts b/arch/arm64/boot/dts/amlogic/meson-gxm-mecool-kiii-pro.dts
deleted file mode 100644
index 453779ac1d14..000000000000
--- a/arch/arm64/boot/dts/amlogic/meson-gxm-mecool-kiii-pro.dts
+++ /dev/null
@@ -1,22 +0,0 @@
-// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
-
-/*
- * Author: Christian Hewitt <christianshewitt@gmail.com>
- */
-
-#include "meson-gxm-q200.dts"
-
-/ {
-        compatible = "mecool,gxm-kiii-pro", "amlogic,s912",
-		"amlogic,meson-gxm";
-	model = "MeCool KIII Pro";
-
-	memory@0 {
-		device_type = "memory";
-		reg = <0x0 0x0 0x0 0xC0000000>;
-	};
-};
-
-&ir {
-        linux,rc-map-name = "rc-mecool-kiii-pro";
-};
