diff --git a/configs/nanopim3_defconfig b/configs/nanopim3_defconfig
index 28a1cb1..528e451 100644
--- a/configs/nanopim3_defconfig
+++ b/configs/nanopim3_defconfig
@@ -11,12 +11,14 @@ CONFIG_FIT=y
 # CONFIG_CMD_FLASH is not set
 # CONFIG_CMD_FPGA is not set
 # CONFIG_CMD_SETEXPR is not set
-# CONFIG_CMD_NET is not set
-# CONFIG_CMD_NFS is not set
+CONFIG_CMD_NET=y
+CONFIG_CMD_NFS=y
+CONFIG_CMD_PXE=y
 CONFIG_CMD_FDISK=y
 CONFIG_CMD_EXT4_IMG_WRITE=y
 CONFIG_CMD_SD_RECOVERY=y
 CONFIG_OF_CONTROL=y
 CONFIG_OF_EMBED=y
+CONFIG_SPL=y
 CONFIG_VIDEO_NX=y
 CONFIG_VIDEO_NX_HDMI=y
diff --git a/configs/s5p6818_drone_defconfig b/configs/s5p6818_drone_defconfig
index 0858a25..b6b707c 100644
--- a/configs/s5p6818_drone_defconfig
+++ b/configs/s5p6818_drone_defconfig
@@ -12,6 +12,7 @@ CONFIG_FIT=y
 # CONFIG_CMD_SETEXPR is not set
 # CONFIG_CMD_NET is not set
 # CONFIG_CMD_NFS is not set
+CONFIG_SPL=y
 CONFIG_CMD_FDISK=y
 CONFIG_CMD_EXT4_IMG_WRITE=y
 CONFIG_CMD_SD_RECOVERY=y
diff --git a/include/configs/s5p6818_nanopim3.h b/include/configs/s5p6818_nanopim3.h
index 11df7e7..f1d4735 100644
--- a/include/configs/s5p6818_nanopim3.h
+++ b/include/configs/s5p6818_nanopim3.h
@@ -100,7 +100,7 @@
  *	U-Boot Environments
  */
 /* refer to common/env_common.c	*/
-#define CONFIG_BOOTDELAY			1
+#define CONFIG_BOOTDELAY			0

 /*-----------------------------------------------------------------------
  * Miscellaneous configurable options
@@ -289,19 +289,37 @@
 /*-----------------------------------------------------------------------
  * ENV
  */
-#define CONFIG_EXTRA_ENV_SETTINGS	\
-			"fdt_high=0xffffffffffffffff\0" \
-			"stdin=serial\0" \
-			"stdout=serial,vga\0" \
-			"stderr=serial,vga"
-
-#define CONFIG_BOOTARGS \
-    "console=ttySAC0,115200n8 initrd=0x49000000,0x2000000"
-
-#define CONFIG_BOOTCOMMAND	\
-	"echo Waiting for environment...;" \
-	"udown 0x48000000; " \
-	"env import -t -r 0x48000000; " \
-	"run bootcmd2"
+#include <config_distro_defaults.h>
+#define BOOT_TARGET_DEVICES(func) \
+	func(MMC, mmc, 1)
+#include <config_distro_bootcmd.h>
+
+#define UDOWN_BOOT \
+	"udown_boot=\
+		echo Waiting for environment...;\
+		udown 0x48000000;\
+		env import -t -r 0x48000000;\
+		echo Running bootcmd2;\
+		run bootcmd2;\0"
+
+#define MEM_LAYOUT_ENV_SETTINGS \
+	"ramdisk_addr_r=0x49000000\0" \
+	"kernel_addr_r=0x48000000\0" \
+	"scriptaddr=0x47000000\0" \
+	"pxefile_addr_r=0x00000000\0" \
+	"fdt_addr_r=0x4a000000\0" \
+	"fdt_high=0xffffffffffffffff\0"
+
+#define CONFIG_EXTRA_ENV_SETTINGS \
+	"fdtfile=nexell/nanopim3.dtb\0" \
+	"stdin=serial\0" \
+	"stdout=serial,vga\0" \
+	"stderr=serial,vga\0" \
+	MEM_LAYOUT_ENV_SETTINGS \
+	UDOWN_BOOT \
+  BOOTENV
+
+#define CONFIG_BOOTCOMMAND \
+	"run distro_bootcmd; echo MMC boot failed, loading from USB; run udown_boot"

 #endif /* __CONFIG_H__ */
